scripts/troubleshoot/cluster-migration/migrate-to-container-insights-msi.sh

```diff
@@ -0,0 +1,333 @@
+#!/bin/bash
+#
+# Script to migrate container insights monitoring to managed identity authentication
+#
+# High Level Steps:
+# 1. Prerequisites Check:
+#    - Validate Azure CLI and login status
+#    - Verify subscription access
+#    - Check cluster identity type
+#    - Verify container insights MSI status
+#    - Check cluster health
+#
+# 2. Cluster Discovery:
+#    - Scan specified subscriptions
+#    - Filter by cluster types (AKS/Arc)
+#    - List eligible clusters
+#
+# 3. Migration Process:
+#    - Get Log Analytics workspace ID
+#    - Disable monitoring addon
+#    - Re-enable with MSI auth
+#
+# Usage:
+#   ./migrate-to-container-insights-msi.sh -s <subscriptionIds> -c <clusterTypes>
+#
+# Example:
+#   ./migrate-to-container-insights-msi.sh -s "sub1,sub2" -c "aks,arc"
+#
+
+# Function to display usage
+usage() {
+    echo "Usage: $0 -s <subscriptionIds> -c <clusterTypes>"
+    echo "  -s : Comma-separated list of subscription IDs"
+    echo "  -c : Comma-separated list of cluster types (aks,arc)"
+    echo "Example:"
+    echo "  $0 -s \"sub1,sub2\" -c \"aks,arc\""
+    exit 1
+}
+
+# Function to check prerequisites for a cluster
+check_prerequisites() {
+    local cluster_type=$1
+    local resource_group=$2
+    local name=$3
+
+    # 1. Check if cluster uses managed identity
+    if [ "$cluster_type" = "aks" ]; then
+        local identity_type=$(az aks show -g "$resource_group" -n "$name" --query "identity.type" -o tsv)
+        if [ "$identity_type" != "SystemAssigned" ] && [ "$identity_type" != "UserAssigned" ]; then
+            echo "Current identity type: $identity_type (requires SystemAssigned or UserAssigned)" >&2
+            echo "To migrate to managed identity, visit: https://learn.microsoft.com/en-us/azure/aks/use-managed-identity" >&2
+            return 1
+        fi
+    fi
+
+    # 2. Check if monitoring is already using MSI
+    if [ "$cluster_type" = "aks" ]; then
+        local auth_mode=$(az aks show -g "$resource_group" -n "$name" --query "addonProfiles.omsagent.config.useAADAuth" -o tsv)
+        if [ "$auth_mode" = "true" ]; then
+            echo "Monitoring already using MSI authentication" >&2
+            return 1
```

Comment:
Lets say, if I have subId1, few clusters in MSI and few clusters in legacy. will the script migrate the legacy clusters to MSI or exists since it found there are already MSI clusters ?

Reply:
it will skip the msi clusters only migrate one with legacy but managed  identity for cluster enablled